### PR TITLE
Yyuen/hash service fix for back button

### PR DESCRIPTION
**Pull request by justYves**
### Changes:
This refactor some of the hashservices.js so that 
- it doesn't rewrite history 
- and so that the browser's back and forward button works

hnki2me6rg

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js [hunk 1]
```diff
@@ -112,10 +112,41 @@ HashService.prototype = {
...
-    this.currentController.handleAppEvent();
+
+  /**
+   * determine based on the hash change event if a transition is needed
+   * @param  {string} options.oldURL [oldURL of hash change event]
+   * @param  {string} options.newURL [newURL of hash change event]
+   * @return {boolean} - returns true if a transition is needed]
+   */
+  shouldTransition({ oldURL, newURL }) {
+    const [_1, transitionFrom, paramsFrom]= oldURL.split(/(?:[#?]|[&]*rand)/);
```
Comment (theseyi) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js]: Might be helpful to add a comment here on what the intention behind matching the charaters `rand` after the quantifier

File: thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js [hunk 2]
```diff
@@ -112,10 +112,41 @@ HashService.prototype = {
...
+    const [_2, transitionTo, paramsTo] = newURL.split(/(?:[#?]|[&]*rand)/);
```
Comment (theseyi) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js]: Same here

File: thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/controllers/DashboardController.js [hunk 3]
```diff
@@ -22,18 +22,22 @@ DashboardController.prototype = {
...
+        controllerName = 'metricSummary';
+        break;
+      case constants.DASHBOARD_MODE_WOW_SUMMARY:
+        this.dashboardModel.tabSelected = 'dashboard_wow-summary-tab';
+        controllerName = 'wowSummary';
+        break;
+      default:
+        this.dashboardModel.tabSelected = 'dashboard_anomaly-summary-tab';
+        controllerName = 'anomalySummary';
+        break;
```
Comment (theseyi) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/controllers/DashboardController.js]: You don't need a `break` statement in a switch's `default` clause as there are no more cases to fall-through

File: thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js [hunk 4]
```diff
@@ -110,12 +110,41 @@ HashService.prototype = {
...
     this.currentController.handleAppEvent();
+  },
+
+  /**
+   * determine based on the hash change event if a transition is needed
+   * @param  {string} options.newURL [newURL of hash change event]
+   * @return {boolean} - returns true if a transition is needed]
+   */
+  shouldTransition({ newURL }) {
+    const [, transitionTo, paramsTo] = newURL.split(/(?:[#?]|[&]*rand)/);
```
Comment (kishoreg) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js]: can you add a simple string that this is trying to split. I possible avoid complex regex. It's ok to break it up into multiple steps.

File: thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js [hunk 5]
```diff
@@ -88,19 +98,18 @@ HashService.prototype = {
...
-        paramsUrl = paramsUrlTokens[1];
-        paramsUrl.split("&").forEach(function(part) {
-          var pair = part.split("=");
-          self.set(pair[0], decodeURIComponent(pair[1]));
+      const [newTab, newParams] = paramsUrl.split('?');
+      if (newParams) {
+        this.set(HASH_PARAMS.TAB, newTab);
+
+        newParams.split("&").forEach((part) => {
+          const [key, value] = part.split("=");
```
Comment (justinpark) [inline thirdeye/thirdeye-pinot/src/main/resources/assets/javascript/HashService.js]: Let's use a single quote.

**Comment by justinpark** [comment]
It looks good. Ship it!

**Pull request merged**